sum_vis and diff_vis methods

ntk688 opened the pull request:
Adds two methods to sum or difference visibilities. Closes #501. 

## Description
Two methods sum_vis and diff_vis added to the UVData object. diff_vis just calls sum_vis with a flag. 

## Motivation and Context
Adds a new feature in response to issue https://github.com/RadioAstronomySoftwareGroup/pyuvdata/issues/501

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation change (documentation changes only)
- [ ] Version change
- [ ] Build or continuous integration change


## Checklist:
- [x] I have read the [contribution guide](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/.github/CONTRIBUTING.md).
- [x] My code follows the code style of this project.

Bug fix checklist:
- [ ] My fix includes a new test that breaks as a result of the bug (if possible).
- [ ] All new and existing tests pass in both python 2 and python 3.
- [ ] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md).

New feature checklist:
- [x] I have added or updated the docstrings associated with my feature using the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [x] I have updated the tutorial to highlight my new feature (if appropriate).
- [x] I have added tests to cover my new feature.
- [x] All new and existing tests pass in both python 2 and python 3.
- [x] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md).

Breaking change checklist:  
- [ ] I have updated the docstrings associated with my change using the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] I have updated the tutorial to reflect my changes (if appropriate).
- [ ] My change includes backwards compatibility and deprecation warnings (if possible).
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests pass in both python 2 and python 3.
- [ ] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md).

Documentation change checklist:
- [ ] Any updated docstrings use the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] If this is a significant change to the readme or other docs, I have checked that they are rendered properly on ReadTheDocs. (you may need help to get this branch to build on RTD, just ask!)

Version change checklist:
- [ ] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md) to put all the unreleased changes under the new version (leaving the unreleased section empty).
- [ ] I have noted any dependency changes since the last version and will update the conda package build accordingly.

Build or continuous integration change checklist:
- [ ] If required or optional dependencies have changed (including version numbers), I have updated the readme to reflect this.
- [ ] If this is a new CI setup, I have added the associated badge to the readme and to references/make_index.py (if appropriate).

bhazelton commented on pyuvdata/uvdata.py:
```diff
@@ -1424,6 +1424,118 @@ def reorder_blts(self, order='time', minor_order=None, conj_convention=None, uvw
+                                '_blt_order', '_channel_width',
+                                '_dut1', '_earth_omega',
+                                '_extra_keywords', '_flag_array',
+                                '_freq_array', '_gst0', '_history',
```
we probably don't want to require that the histories are the same.

ntk688 replied:
I removed the history parameter. To make this list I just grabbed every parameter, minus the data_array and now the history, so let me know if others are unnecessary.

bhazelton replied:
I think this is a good set for now. Maybe add a comment above this list to say it's everything except the history and the data_array? Or better yet, make the list this way:
```
compatibility_params = list(temp.__iter__())
compatibility_params.remove('_data_array')
compatibility_params.remove('_history')
```

bhazelton commented:
There are special keywords on github that will close the related issue when this PR is merged if you use them properly. It has to go in the initial (top) PR description and it needs to follow the pattern `closes #___` or `fixes #___`.

bhazelton commented:
I just enabled circleci builds for pull requests from forks (I didn't realize it wasn't enabled until now), so when you push another commit it should actually trigger the circleci builds.

mkolopanis commented:
This will need to be rebased for those github actions to pass now.

mkolopanis commented:
ah the hera_qm failure here is my fault from a QM pr we just merged in. I offloaded a lot of command line arguments to the setup.cfg but that basically requires pytest-cov to run tests. that could be fixed by adding `pytest-cov` to the hera_qm environemnt yaml.

hera_cal is a bit confusing though. It is not downloading `uvtools` anymore via pip? @adampbeardsley do you know if something recently changed in hera_cal that would affect this?

ntk688 commented:
Ok I rebased this, and it now has an updated changelog and a brief demonstration in the "adding data" section of the tutorial.

adampbeardsley commented:
There's this:
https://github.com/HERA-Team/hera_cal/commit/03a1fee8228b21bcf4d0242be524293eabed34d8
But that was three months ago.
The only recent change I know of is that Josh changed the environment to use the most recent pyuvdata. But I don't think that should matter here.

mkolopanis commented:
Ah i think it was a uvtools thing. I made a note of it in #728. The workaround is to manually install basemap because pip is having trouble finding it now? Thanks for digging though.

bhazelton commented:
I'm not sure why the coverage isn't reporting, must have to do with this being a fork, but I can't find a place to fix it in the settings on codecov. I downloaded your branch and ran `test_cover_noinstall.sh` from the scripts folder, which makes output similar to codecov and I found that there are 4 new uncovered lines, line numbers 1464, 1487, 1489 and 1515. The first and last of those will only get called for an inlace sum or diff, the middle two are for a ValueError, we test that Errors are raised as expected using something like this (stolen from another test, you'll need to update for your error):
```    with pytest.raises(ValueError) as cm:
        uvdata_data.uv_object.check()
    assert str(cm.value).startswith('All antennas in ant_1_array must be in antenna_numbers')
```

ntk688 commented:
I had some of those error tests, but when I went to investigate I realized those lines weren't being called because of a goofy indenting mistake. So I fixed that and added the in place checks and now test_cover_noinstall.sh shows no uncovered lines in the code I added. Codecov still not going through though...
